plugin/advisor/mysql/advisor_index_pk_type.go

```diff
@@ -0,0 +1,276 @@
+package mysql
+
+// Framework code is generated by the generator.
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/parser/types"
+	"github.com/pkg/errors"
+
+	"github.com/bytebase/bytebase/plugin/advisor"
+	"github.com/bytebase/bytebase/plugin/advisor/catalog"
+	"github.com/bytebase/bytebase/plugin/advisor/db"
+)
+
+var (
+	_ advisor.Advisor = (*IndexPkTypeAdvisor)(nil)
+	_ ast.Visitor     = (*indexPkTypeChecker)(nil)
+)
+
+func init() {
+	advisor.Register(db.MySQL, advisor.MySQLIndexPKType, &IndexPkTypeAdvisor{})
+	advisor.Register(db.TiDB, advisor.MySQLIndexPKType, &IndexPkTypeAdvisor{})
+}
+
+// IndexPkTypeAdvisor is the advisor checking for correct type of PK.
+type IndexPkTypeAdvisor struct {
+}
+
+// Check checks for correct type of PK.
+func (*IndexPkTypeAdvisor) Check(ctx advisor.Context, statement string) ([]advisor.Advice, error) {
+	stmtList, errAdvice := parseStatement(statement, ctx.Charset, ctx.Collation)
+	if errAdvice != nil {
+		return errAdvice, nil
+	}
+
+	level, err := advisor.NewStatusBySQLReviewRuleLevel(ctx.Rule.Level)
+	if err != nil {
+		return nil, err
+	}
+	checker := &indexPkTypeChecker{
+		level:            level,
+		title:            string(ctx.Rule.Type),
+		line:             make(map[string]int),
+		database:         ctx.Database,
+		tablesNewColumns: make(map[string]columnNameToColumnDef),
+	}
+
+	for _, stmt := range stmtList {
+		(stmt).Accept(checker)
+	}
+
+	if len(checker.adviceList) == 0 {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  advisor.Success,
+			Code:    advisor.Ok,
+			Title:   "OK",
+			Content: "",
+		})
+	}
+	return checker.adviceList, nil
+}
+
+type columnNameToColumnDef map[string]*ast.ColumnDef
+type tableNewColumn map[string]columnNameToColumnDef
+
+func (t tableNewColumn) set(tableName string, columnName string, colDef *ast.ColumnDef) {
+	if _, ok := t[tableName]; !ok {
+		t[tableName] = make(map[string]*ast.ColumnDef)
+	}
+	t[tableName][columnName] = colDef
+}
+
+func (t tableNewColumn) get(tableName string, columnName string) (colDef *ast.ColumnDef, ok bool) {
+	if _, ok := t[tableName]; !ok {
+		return nil, false
+	}
+	col, ok := t[tableName][columnName]
+	return col, ok
+}
+
+func (t tableNewColumn) delete(tableName string, columnName string) {
+	if _, ok := t[tableName]; !ok {
+		return
+	}
+	delete(t[tableName], columnName)
+}
+
+type indexPkTypeChecker struct {
+	adviceList       []advisor.Advice
+	level            advisor.Status
+	title            string
+	line             map[string]int
+	database         *catalog.Database
+	tablesNewColumns tableNewColumn
+}
+
+type pkData struct {
+	table      string
+	column     []string
+	columnType []string
+	line       int
+}
```

Comment:
Please refactor this struct and related code.
```suggestion
type pkData struct {
	table      string
	column     string
	columnType string
	line       int
}
```